server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java

```diff
@@ -0,0 +1,282 @@
+import org.infinispan.marshall.protostream.impl.SerializationContextRegistry;
+import org.infinispan.metadata.Metadata;
+import org.infinispan.metadata.impl.InternalMetadataImpl;
+import org.infinispan.metadata.impl.PrivateMetadata;
+import org.infinispan.protostream.ImmutableSerializationContext;
+import org.infinispan.protostream.annotations.ProtoField;
+import org.infinispan.protostream.annotations.ProtoTypeId;
+import org.infinispan.reactive.publisher.PublisherTransformers;
+import org.infinispan.reactive.publisher.impl.ClusterPublisherManager;
+import org.infinispan.reactive.publisher.impl.DeliveryGuarantee;
+import org.infinispan.registry.InternalCacheRegistry;
+import org.infinispan.server.core.BackupManager;
+import org.infinispan.util.concurrent.AggregateCompletionStage;
+import org.infinispan.util.concurrent.BlockingManager;
+import org.infinispan.util.concurrent.CompletionStages;
+import org.reactivestreams.Publisher;
+
+import io.reactivex.rxjava3.core.Flowable;
+
+/**
+ * {@link org.infinispan.server.core.backup.ContainerResource} implementation for {@link
+ * BackupManager.Resources.Type#CACHES}.
+ *
+ * @author Ryan Emerson
+ * @since 12.0
+ */
+public class CacheResource extends AbstractContainerResource {
+
+   private static final String MEMCACHED_CACHE = "memcachedCache";
+
+   // TODO what size?
+   private static final int BUFFER_SIZE = 100;
+
+   private final EmbeddedCacheManager cm;
+   private final ParserRegistry parserRegistry;
+
+   CacheResource(BlockingManager blockingManager, ParserRegistry parserRegistry, EmbeddedCacheManager cm,
+                 BackupManager.Resources params, Path root) {
+      super(CACHES, params, blockingManager, root);
+      this.cm = cm;
+      this.parserRegistry = parserRegistry;
+   }
+
+   @Override
+   public void prepareAndValidateBackup() {
+      InternalCacheRegistry icr = cm.getGlobalComponentRegistry().getComponent(InternalCacheRegistry.class);
+
+      Set<String> caches = wildcard ? cm.getCacheConfigurationNames() : resources;
+      for (String cache : caches) {
+         Configuration config = cm.getCacheConfiguration(cache);
+
+         if (wildcard) {
+            // For wildcard resources, we ignore internal caches, however explicitly requested internal caches are allowed
+            if (config == null || config.isTemplate() || icr.isInternalCache(cache) || MEMCACHED_CACHE.equals(cache)) {
+               continue;
+            }
+            resources.add(cache);
+         } else if (config == null) {
+            throw log.unableToFindResource(type.toString(), cache);
+         } else if (config.isTemplate()) {
+            throw new CacheException(String.format("Unable to backup %s '%s' as it is a template not a cache", type, cache));
+         }
+      }
+   }
+
+   @Override
+   public CompletionStage<Void> backup() {
+      AggregateCompletionStage<Void> stages = CompletionStages.aggregateCompletionStage();
+      for (String cache : resources)
```

Comment:
It makes me wonder if we should add some extra methods to `BlockingManager#BlockingExecutor` to handle these types of cases.

Reply:
Agreed that we should limit the number of blocking threads are used by this, as even if there is a reasonably modest number of caches this could cause issues.

I think `BlockingManager.BlockingExecutor` should provide all of the `BlockingManager` methods with the exception of `limitedBlockingExecutor`. That way we can do something like:

```java
      BlockingManager.BlockingExecutor executor = blockingManager.limitedBlockingExecutor("cache-backup-executor", 5);
     for (String cache : resources)
         stages.dependsOn(createCacheBackup(cache, executor));
```

To reduce repetition in the `BlockingManager` interface, we can create another interface `BlockingOperations` and then do the following:

```java
public interface BlockingManager extends BlockingOperations {

    BlockingExecutor limitedBlockingExecutor(String name, int concurrency);

    interface BlockingExecutor extends BlockingOperations {
    }
}
```

`BlockingManagerImpl` would then just provide versions of the `BlockingOperations` methods that take a `Scheduler|Executor` as an additional parameter.

Reply:
Similarly we could avoid `BlockingOperations` and add a method that returns a BackupManager implementations that utilies a `LimitedExecutor`:

```java
BlockingManager limitedBlockingManager(String name, int concurrency);
```

Reply:
I look forward to seeing this change in another PR 👍

Reply:
I have created [ISPN-12226](https://issues.redhat.com/browse/ISPN-12226) to track the `BlockingManager` enhancements and [ISPN-12227](https://issues.redhat.com/browse/ISPN-12227) for the changes to the `BackupManager`.